fix: track JetBrains connections

code-asher opened the pull request:
Fixes #9673 

Possibly we will need to find a better string to match on, but this one seems OK to me.

Unit test passes but have not tested an actual deploy yet.

Emyrk committed:
feat: implement jetbrains agentssh tracking

Based on tcp forwarding instead of ssh connections

Emyrk committed:
Add unit test to confirm tracking

Emyrk committed:
implement unit test to verify jetbrains functionality

code-asher committed:
Implement port process inspection

code-asher committed:
Add JetBrains tracking to bottom bar

code-asher commented:
Ah wait moving back to draft, there seems to be an issue...

Emyrk commented on agent/agent_test.go:
```diff
@@ -191,6 +191,60 @@ func TestAgent_Stats_Magic(t *testing.T) {
 		require.NoError(t, err)
 	})
+
+	// This test name must contain the string checked for by the agent, since it
+	// looks for this string in the process name.
+	t.Run("TracksIdea.vendor.name=JetBrains", func(t *testing.T) {
```
Lets add some more comments as this is not intuitive. 

```suggestion
	// This test name must contain the string checked for by the agent, since it
	// looks for this string in the process name.
	//
	// This test sets up a port forward that emulates what Jetbrains IDE's do when
	// using gateway. The remote server side of the port forward is spun up using
	// the gotest process, which includes the test name.
	// So this unit test emulates a PID in the workspace with a similar
	// name to the jetbrains IDE. That makes the agent this this SSH port
	// forward is a "jetbrains" session.
	t.Run("TracksIdea.vendor.name=JetBrains", func(t *testing.T) {
```

Emyrk commented on agent/agentssh/jetbrainstrack.go:
```diff
@@ -0,0 +1,85 @@
+	gossh "golang.org/x/crypto/ssh"
+)
+
+type localForwardChannelData struct {
```
```suggestion
// localForwardChannelData is copied from the ssh package.
type localForwardChannelData struct {
```

Emyrk commented on agent/agentssh/jetbrainstrack.go:
```diff
@@ -0,0 +1,85 @@
+	OriginPort uint32
+}
+
+type ChannelAcceptWatcher struct {
```
```suggestion
// ChannelAcceptWatcher is used to track jetbrains port forwarding (gateway)
// connections. If the port forward is something other than jetbrains, this
// struct is a noop.
type ChannelAcceptWatcher struct {
```

Emyrk replied:
Lets call it something pretty direct?
```suggestion
type JetbrainsNewChannel struct {
```

Emyrk commented on agent/agentssh/jetbrainstrack.go:
```diff
@@ -0,0 +1,85 @@
+	jetbrainsCounter *atomic.Int64
+}
+
+func NewChannelAcceptWatcher(ctx ssh.Context, logger slog.Logger, newChannel gossh.NewChannel, counter *atomic.Int64) gossh.NewChannel {
```
Is this a bad name?
```suggestion
func WatchForJetbrains(ctx ssh.Context, logger slog.Logger, newChannel gossh.NewChannel, counter *atomic.Int64) gossh.NewChannel {
```

code-asher replied:
Yeah I think you are right, will rename!

code-asher committed:
Elaborate on process name check comment

Co-authored-by: Steven Masley <Emyrk@users.noreply.github.com>

code-asher committed:
Comment that localForwardChannelData is copied

Co-authored-by: Steven Masley <Emyrk@users.noreply.github.com>

code-asher committed:
Comment ChannelAccepterWatcher

Co-authored-by: Steven Masley <Emyrk@users.noreply.github.com>

code-asher committed:
Rename channel watcher to be specific to Jetbrains

code-asher commented:
The two issues I ran into, just FYI:

- I had the process name check reversed, needed a `!` :facepalm: 
- Netstat only gives the base name, not the full command line, so I need to read `/proc/$pid/cmdline` to get that.

I think it is all sorted now, doing some more testing to be sure.

code-asher committed:
Log unmarshal failure

matifali commented:
Will this also work for Fleet?

code-asher committed:
Fix JetBrains tracking test

The test name only shows up in the process name if you are running that
test directly so we have to spawn a separate process instead.

code-asher commented:
> Will this also work for Fleet?

I am not sure, we will need to test.

code-asher committed:
Add constant for JetBrains magic string

Emyrk reviewed (approved):
Looks good, let's get this in and test it a bit.
